### ISSUE TITLE
No Spotify artwork in notifications and Raven

### ISSUE THREAD

**Issue by tristan957** [budgie-desktop/budgie-desktop#560]
Artwork seems to be the only thing missing from creating a seamless Spotify integration.

**Comment by ikeydoherty** [comment]
Its a Spotify issue in reality, they use remote URLs for the artwork locations.
To support it would require pimping Raven significantly to permit fetches, and then we have to consider bandwidth and such things. Perhaps some option to enable/disable it.. Thoughts?

**Comment by tristan957** [comment]
FYI I'm a sophomore in college and I know Java really well just started learning C++ as well as Python. So my ideas will probably suck.

My first idea would be to see what the Cinnamon and Ubuntu music applets are doing to display artwork. As for the bandwidth problem, maybe their could be code in the Raven music applet like if Spotify is running, display an option within the applet to disable fetching the album art. The choice could be persistent across sessions until the user changes the option. The option could be hidden as a transparent gear icon that sits above the album art in the top right. I have no idea if this is feasible or not because I don't know the limitations of Raven/whatever language Raven is written in.

**Comment by KeizerDev** [comment]
@tristan957 you should write a plugin for it like https://extensions.gnome.org/extension/55/media-player-indicator/ wouldn't be that hard to create something like that if you follow [this](https://xpressubuntu.wordpress.com/2016/05/07/writing-python-applets-for-the-budgie-desktop/) guide.

You could even create a plugin that downloads it for you and pipes it to the default applet ( correct me if i'm wrong @ikeydoherty ).

**Comment by ikeydoherty** [comment]
Pls don't use Python. lol.

**Comment by tristan957** [comment]
@ikeydoherty what is wrong with Python? I thought it was a popular language. Thoughts on a better language to learn instead? Trying to make myself marketable to employers.

**Comment by ikeydoherty** [comment]
I'm talking about for Budgie applets. It's a bastard to debug the runtime issues and at least a compiled language will protect you to some extent by  checking errors in advance.

**Comment by tristan957** [comment]
Ahh. Not saying I would code it in Python. I was just providing background as to how new I am at this. If I find some time I will look into possible solutions.

**Comment by mussel** [comment]
@ikeydoherty , I saw that Raven uses org.freedesktop.Notifications to notify the user of a new song being played by Spotify, for instance. But I could not see which entity issues the Notify command. Is it MprisGui or some other object belonging to raven or is it Spotify directly?

**Comment by ikeydoherty** [comment]
@mussel that's not **quite** the way it works :) Spotify calls out to `org.freedesktop.Notifications` to send a notification. Raven provides a service on `org.freedesktop.Notifications` to **receive** notifications and act as the notification server for the desktop.

Spotify itself is the one sending things

**Comment by mussel** [comment]
Thanks for clarifying that @ikeydoherty 

So I did some work on this issue and this is the result:

![cover_art_support](https://cloud.githubusercontent.com/assets/1262461/19748388/1e90f99a-9bc0-11e6-9a88-70452332e32c.png)


So, I had to add support for two things
**Download cover art**
I created an async function, in MprisGui.vala, that creates a stream to a file in the internet and then uses the a Pixbuf to load the image from the stream, asynchronously. With this I was able to add the cover art to the Raven wiget.

**Load images sent through DBUS**
To add support for images in the notifications, I created a function that loads images(iiibiiay) sent as hints of a notification. Moreover, I also changed the order in which NotificationView looks for images to be compliant with the org.freedesktop.Notifications spec.

Does this sound reasonable?

So I wanted to clarify two more things before I open a PR:
1. As you can see in the screenshot, there is some space on the sides of the cover art in the Raven widget. Was it intended to be this way or is this mismatch in the proportions of the image and the widget?
2. I noticed that when skipping to the next/prev track, MprisGui receives **4** PropertiesChanged notifications from DBUS. Were you aware of this Ikey? I did, however, put a check in my function to only download the new covert art if its url is different from the current one, to prevent unnecessary downloads.

**Comment by ikeydoherty** [comment]
Notice how we load images at a specific size + scale - those would be the ones to investigate. Then just resizing your image to fit.

As for the properties changed, it all depends on how buggy/well-coded the client is. It hasn't been a problem so far as we don't do any remote fetching.

In your fetching implementation, are you handling timeouts and cancellation of previously running fetch tasks? How well does it handle if you flick through rapidly in the playlist? Is there also a way to switch this off?

**Comment by mussel** [comment]
## Cover art size
So, I already using the Pixbuf to scale the image:
```vala
Gdk.Pixbuf? pbuf = yield new Gdk.Pixbuf.from_stream_at_scale_async(ins, this.our_width,
this.our_width,  true, cancel);
```
As you can see, this makes a square image with the same dimensions as the ClientWidget.

#### MprisWidget Bug
So I did some digging on Raven's code and found the issue. When *MprisWidget* creates
a new *ClientWidget* for a player registered in DBUS, it passes to the *ClientWidget*'s
ctor the size indicated by the variable *our_width*. Initially, this variable
has value 250, but once the *MprisWidget* is realized, it receives the actual
width of the widget. However, if a player, e.g. Spotify, is opened **before**
Raven is opened by the first time, then the *MprisWidget* will not be realized and
it will pass the width 250 to *ClientWidget*. And here is the issue: Once raven
is opened and *MprisWidget* is realized, it does not pass to *ClientWidget* its new
size.

##### Solution
So there are two possible solutions to this bug in specific
1. *MprisWidget* *on_size_allocate* function updates all *ClientWidgets* on their
new size
2. *ClientWidget* instances connect to the *size_allocate* signal themselves
and auto-update their size

In addition to implementing one of the solutions above, *ClientWidget* must
also be able to resize the cover art image being displayed, once its size is
updated. This way, IF you start listening to a song in Spotify before opening Raven,
when you do open it, *ClientWidget* will make the cover art the appropriate size.

**So, which one of the solutions above do you think is more appropriate?**




## Fetching implementation
My implementation does cancel already pending fetches but I am not sure how
to deal with timeouts. The code that I used is very similar to the one you wrote
to load images from a path asynchronously:
```vala
async void download_art(string uri)
    {
        try {
            // open the stream
            var art_file = GLib.File.new_for_uri (uri);
            // download the art
            var ins = yield art_file.read_async(Priority.DEFAULT, cancel);
            Gdk.Pixbuf? pbuf = yield new Gdk.Pixbuf.from_stream_at_scale_async(ins,
                this.our_width, this.our_width, true, cancel);
            background.set_from_pixbuf(pbuf);
        } catch (Error e) {
            update_art_fallback();
        }
    }
```
Here, I use the *Cancellable* instance *cancel* to cancel pending fetches. Furthermore,
as I mentioned in my previous post, repeated metadata updates are not an issue
for my implementation, since it only fetches the cover art if the new URI is
different from the current one.


Lastly, I did test quickly flick with different covert arts and from a performance
standpoint it worked fine. The only thing is that after a couple skips
the covert art fallbacks to the "emblem-music-symbolic" icon, while the cover art
is being fetched. Once the download finishes, the cover art is updated accordingly.


## Summary
1. Which solution to the bug do you think would work best?
2. Once I fix this bug, I will create a PR and we can continue the discussion there.

**Comment by tristan957** [comment]
Real life Sherlock Holmes right here.

**Comment by m-delvalle** [comment]
@mussel too sad there's no "when I'm grown up, I want to be like you" badge   ;)

**Comment by ikeydoherty** [comment]
Fixed in upcoming 10.3
